src/java.base/share/classes/java/lang/StableValue.java

```diff
@@ -0,0 +1,664 @@
+ * stable value storage for that parameter value. A stable function is created via the
+ * {@linkplain StableValue#function(Set, Function) StableValue.function()} factory.
+ * Upon creation, the input {@linkplain Set} is specified together with an original
+ * {@linkplain Function} which is invoked at most once per input value. In effect, the
+ * stable function will act like a cache for the original {@linkplain Function}:
+ *
+ * {@snippet lang = java:
+ * public final class SqrtUtil {
+ *
+ *     private SqrtUtil(){}
+ *
+ *     private static final Function<Integer, Double> SQRT =
+ *             // @link substring="function" target="#function(Set,Function)" :
+ *             StableValue.function(Set.of(1, 2, 4, 8, 16, 32), StrictMath::sqrt);
+ *
+ *     public static double sqrt16() {
+ *         return SQRT.apply(16); // May eventually constant fold to 4.0 at runtime
+ *     }
+ *
+ * }
+ *}
+ *
+ * <h2 id="stable-collections">Stable Collections</h2>
+ * Stable values can also be used as backing storage for
+ * {@linkplain Collection##unmodifiable unmodifiable collections}. A <em>stable list</em>
+ * is an unmodifiable list, backed by an array of stable values. The stable list elements
+ * are computed when they are first accessed, using a provided {@linkplain IntFunction}:
+ *
+ * {@snippet lang = java:
+ * public final class SqrtUtil {
+ *
+ *     private SqrtUtil(){}
+ *
+ *     private static final List<Double> SQRT =
+ *             // @link substring="list" target="#list(int,IntFunction)" :
+ *             StableValue.list(10, StrictMath::sqrt);
+ *
+ *     public static double sqrt9() {
+ *         return SQRT.get(9); // May eventually constant fold to 3.0 at runtime
+ *     }
+ *
+ * }
+ *}
+ * <p>
+ * Similarly, a <em>stable map</em> is an unmodifiable map whose keys are known at
+ * construction. The stable map values are computed when they are first accessed,
+ * using a provided {@linkplain Function}:
+ *
+ * {@snippet lang = java:
+ * public final class SqrtUtil {
+ *
+ *     private SqrtUtil(){}
+ *
+ *     private static final Map<Integer, Double> SQRT =
+ *             // @link substring="map" target="#map(Set,Function)" :
+ *             StableValue.map(Set.of(1, 2, 4, 8, 16, 32), StrictMath::sqrt);
+ *
+ *     public static double sqrt16() {
+ *         return SQRT.get(16); // May eventually constant fold to 4.0 at runtime
+ *     }
+ *
+ * }
+ *}
+ *
+ * <h2 id="composition">Composing stable values</h2>
+ * A stable value can depend on other stable values, thereby creating a dependency graph
+ * that can be lazily computed but where access to individual elements still can be
+ * constant-folded. In the following example, a single {@code Foo} and a {@code Bar}
+ * instance (that is dependent on the {@code Foo} instance) are lazily created, both of
+ * which are held by stable values:
+ * {@snippet lang = java:
+ * public final class DependencyUtil {
```

Comment:
I think the fibonacci example is great -- and is probably enough for this section. We also had, at some point, a nice ASCII art of the computations involved in the fibonacci, which might be useful to illustrate your claim re. "dependency graph". Do you think you can recover that ?

Reply:
I found it! Soon to be added!